Update project tags type to what is expected/returned by API

dsciacca opened the pull request:
# Description

This PR updates the type hints of the `tags` property of the `Project` class to align with what is expected and returned by the Doccano API. This addresses issues around creating, updating, and retrieving Doccano projects that have tags associated. Now that the issue mentioned in #149 is addressed I've found that the `DoccanoClient.update_project()` method fails to actually update tags when that kwarg is passed. Will open a separate issue for this since you can't actually reproduce the failed update of tags unless this PRs change is part of your local `doccano-client`

Fixes #149 (Unable to retrieve projects with tags)

Reproduction of issue before fix:
`find_project_by_id()`:
![Screenshot 2023-06-05 at 9 19 33 AM](https://github.com/doccano/doccano-client/assets/39314769/45d26cb9-6068-4a8e-9533-1bc03bc7bed5)
`list_projects()`:
![Screenshot 2023-06-05 at 9 19 03 AM](https://github.com/doccano/doccano-client/assets/39314769/fb540dbd-05c9-45c4-a150-67f07102aa4e)
`create_project()` w/ tags as specified in [docs](https://doccano.github.io/doccano-client/usage/#doccano_client.client.DoccanoClient.update_project):
![Screenshot 2023-06-05 at 9 17 59 AM](https://github.com/doccano/doccano-client/assets/39314769/92d9cdd0-f84f-4556-a858-602acd893416)
`create_project()` w/ tags as `List[Dict[str, str]]`:
![Screenshot 2023-06-05 at 9 17 11 AM](https://github.com/doccano/doccano-client/assets/39314769/5b43870e-1a12-4443-b96f-b8d245a4b776)
`update_project()` w/ tags as specified in [docs](https://doccano.github.io/doccano-client/usage/#doccano_client.client.DoccanoClient.update_project):
![Screenshot 2023-06-05 at 9 16 03 AM](https://github.com/doccano/doccano-client/assets/39314769/cead0d35-7608-492a-a8a6-3b24b3f23a22)
`update_project()` w/ tags as `List[Dict[str, str]]`:
![Screenshot 2023-06-05 at 9 16 22 AM](https://github.com/doccano/doccano-client/assets/39314769/9f261623-997c-485b-974f-af63665cf007)

Demonstration of successful fix using client created from this branch:
`find_project_by_id()`:
![Screenshot 2023-06-05 at 9 22 12 AM](https://github.com/doccano/doccano-client/assets/39314769/18aa291d-5f6f-479d-9fce-522196a898e7)
`list_projects()` (truncated but demonstrates no failure):
![Screenshot 2023-06-05 at 9 23 16 AM](https://github.com/doccano/doccano-client/assets/39314769/50de58dc-d044-425e-b186-ecf492aef0e3)
`create_project()`:
![Screenshot 2023-06-05 at 9 25 07 AM](https://github.com/doccano/doccano-client/assets/39314769/ccb70ed1-0cd6-4e14-9e96-da9f6ccb9147)
`update_project()` here you'll notice tags aren't actually updated but we can tell the call succeeded as the description was updated, this seems to be a separate bug only reproducible once this PRs fix is in place:
![Screenshot 2023-06-05 at 9 33 47 AM](https://github.com/doccano/doccano-client/assets/39314769/6e0741ff-e029-4b5e-93c5-874172463973)

NOTE: This work was done as part of my employment at [GoGuardian, Inc.](https://github.com/goguardian)

dsciacca commented on doccano_client/client.py:
```diff
@@ -357,7 +357,7 @@ def create_project(
         grapheme_mode: bool = False,
         use_relation: bool = False,
-        tags: Optional[List[str]] = None,
+        tags: Optional[List[Dict[str, str]]] = None,
```
In create/update methods I set this to `Optional[List[Dict[str, str]]]` b/c as far as I can tell the accepted format for a tag in these calls is simply `[{"text", "<tag_name>"}, ...]` and am unsure if it would be accepted by the API to pass any keys aside from `"text"` (i.e. `id` or `project`) in the `tags` dictionaries. Happy to change this to `Optional[List[Dict[str, Any]]]` if it's felt this would be more appropriate

dsciacca commented on doccano_client/models/project.py:
```diff
@@ -47,7 +47,7 @@ class Project(BaseModel):
     grapheme_mode: bool = False
     use_relation: bool = False
-    tags: List[str] = []
+    tags: List[DictStrAny] = []
```
This is the only place w/ `List[Dict[str, Any]]` because when the API returns tags, it includes integer values for the `project` and `id` fields.

Hironsan commented:
Can this problem be fixed by adding and using the following two methods in [ProjectRepository](https://github.com/doccano/doccano-client/blob/master/doccano_client/repositories/project.py#L9)? I think this does not change the signature.

```python
def to_domain(response: dict[str, Any]) -> Project:
    """Convert a response to a domain object

    Args:
        response (dict[str, Any]): The response to convert

    Returns:
        Project: The converted project
    """
    response["tags"] = [tag["text"] for tag in response["tags"]]
    return Project.parse_obj(response)


def to_persistent(project: Project) -> dict[str, Any]:
    """Convert a domain object to a persistent object

    Args:
        project (Project): The project to convert

    Returns:
        dict[str, Any]: The converted project
    """
    payload = project.dict()
    payload["tags"] = [{"text": tag} for tag in payload["tags"]]
    return payload

...
    def find_by_id(self, project_id: int) -> Project:
        response = self._client.get(f"projects/{project_id}")
        return to_domain(response.json())

    def create(self, project: Project) -> Project:
        payload = to_persistent(project)
        response = self._client.post("projects", json=payload)
        return to_domain(response.json())
```

dsciacca commented:
@Hironsan good call, this also works and I agree is a better solution. I've reverted my original change and implemented your suggestion w/ some minor changes (i.e. make functions private class methods and remove "id" from the creation dict, if present) and align w/ expectations of unit tests. Please take a look and let me know your thoughts. Thanks!

Hironsan commented:
Thanks! Merged.
